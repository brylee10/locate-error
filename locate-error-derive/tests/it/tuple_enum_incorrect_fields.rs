//! Enums which use the `Locate` should have at least one variant with a field marked with the #[locate_from] attribute.
//! The fields which have this attribute should be a two element tuple, with the first element being the source and the second element being the location.

use locate_error::Locate;
use locate_error::Location;

// Too many fields
#[derive(Locate)]
enum TestEnum {
    Variant1(#[locate_from] String, Location, String),
}

// No location field
#[derive(Locate)]
enum TestEnum2 {
    Variant1(#[locate_from] String),
}

// No locate_from attribute
#[derive(Locate)]
enum TestEnum3 {
    Variant1(String, Location),
    Variant2(String),
}

// Too many fields with locate_from attribute
#[derive(Locate)]
enum TestEnum4 {
    Variant1(#[locate_from] String, Location, #[locate_from] u32),
}

// No location field
#[derive(Locate)]
enum TestEnum5 {
    Variant1(#[locate_from] String, u32),
}

// Error location should be on the variant name, not the attribute
#[derive(Debug, Locate, thiserror::Error)]
enum TestEnum6 {
    #[error(transparent)]
    Variant1(#[locate_from] SomeError),
}

#[derive(Debug, thiserror::Error)]
#[error("Error message")]
struct SomeError {
    message: String,
}

fn main() {}
