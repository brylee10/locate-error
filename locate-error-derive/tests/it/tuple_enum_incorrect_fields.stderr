error: Locate requires enums variants with the #[locate_from] attribute to have exactly two fields, one for the source and one for the location
  --> tests/it/tuple_enum_incorrect_fields.rs:10:5
   |
10 |     Variant1(#[locate_from] String, Location, String),
   |     ^^^^^^^^

error: Locate requires enums variants with the #[locate_from] attribute to have exactly two fields, one for the source and one for the location
  --> tests/it/tuple_enum_incorrect_fields.rs:16:5
   |
16 |     Variant1(#[locate_from] String),
   |     ^^^^^^^^

error: Locate requires at least one variant with the #[locate_from] attribute (otherwise this macro is effectively a no-op)
  --> tests/it/tuple_enum_incorrect_fields.rs:20:10
   |
20 | #[derive(Locate)]
   |          ^^^^^^
   |
   = note: this error originates in the derive macro `Locate` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Locate requires enums variants with the #[locate_from] attribute to have exactly two fields, one for the source and one for the location
  --> tests/it/tuple_enum_incorrect_fields.rs:29:5
   |
29 |     Variant1(#[locate_from] String, Location, #[locate_from] u32),
   |     ^^^^^^^^

error: Variants with #[locate_from] must have a field of type `locate_from::Location`
  --> tests/it/tuple_enum_incorrect_fields.rs:33:10
   |
33 | #[derive(Locate)]
   |          ^^^^^^
   |
   = note: this error originates in the derive macro `Locate` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Locate requires enums variants with the #[locate_from] attribute to have exactly two fields, one for the source and one for the location
  --> tests/it/tuple_enum_incorrect_fields.rs:42:5
   |
42 |     Variant1(#[locate_from] SomeError),
   |     ^^^^^^^^
